ops/framework.py

```diff
@@ -330,6 +330,21 @@ class Object(metaclass=_Metaclass):
     handle_kind = HandleKind()
 
     def __init__(self, parent, key):
+        """Initialize the Object, tracking the framework reference and the object's unique name.
+
+        Args:
+            parent: The class that this inherits from. Used here to track and maintain a
```

Comment:
Less the 'class' and more the 'instance' I believe.
So not really class-based-inheritence, but a tree-based objects

It also gets a little messy because of meta classes and attributes.

If you look at:
CharmEvents, which inherits from ObjectEvents which is an Object. It uses a handle kind of 'on'.
But then if you look at CharmBase, you have:
```
class CharmBase(Object):
...
    on = CharmEvents()
```
In those cases, there is a presumption of being a singleton in the framework, and the key is None. It uses the magic of `_Metaclass`, and ObjectEvents.__get__ to populate the initialization.

And then if you look at what kind ends up being, it is generally the name of the class or of the attribute on a type. so:
```
harness.charm.handle.path
'CharmBase'
harness.charm.on.handle.path
'CharmBase/on'
```

I realize I'm probably not adding clarity here, but I'd like us to be getting a little bit closer to how it actually works.

Reply:
```
Args:
    parent: The node which this leaf node is created under. This object should be a :class:`Framework` instance or 
                another instance of :class:`Object`. This argument is used to track and signal an object by reference when 
                events are emitted, for :class:`StoredData`, and cache lookups.
    key: the identifier for this node in the reference tree
  ```
  
I don't know that this is any clearer, but it's the way I've interpreted the operating model.
  
`_Metaclass` will die when we can drop support for 3.5, so it's maybe not worth documenting that case only to remove it later. Documenting how events flow in and out of `define_event`, `observe`, and `emit` is definitely worth documenting. I'm not sure that `Object` is the right place to do it compared to `Framework`, though.